File: ios/Classes/AppodealRewarded.swift
import Flutter
import Foundation
import Appodeal

internal final class AppodealRewarded {
    
    internal let adChannel: FlutterMethodChannel
    internal let adListener: Listener
    
    init(registrar: FlutterPluginRegistrar) {
        adChannel = FlutterMethodChannel(name: "appodeal_flutter/rewarded", binaryMessenger: registrar.messenger())
        adListener = Listener(adChannel: adChannel)
    }
    
    internal final class Listener: NSObject, AppodealRewardedVideoDelegate {
        
        private let adChannel: FlutterMethodChannel
        
        fileprivate init(adChannel: FlutterMethodChannel) {
            self.adChannel = adChannel
        }
        
        func rewardedVideoDidLoadAdIsPrecache(_ precache: Bool) {
            adChannel.invokeMethod("onRewardedVideoLoaded", arguments: ["isPrecache": precache])
        }
        
        func rewardedVideoDidFailToLoadAd() {
            adChannel.invokeMethod("onRewardedVideoFailedToLoad", arguments: nil)
        }
        
        func rewardedVideoDidFailToPresentWithError(_ error: Error) {
            adChannel.invokeMethod("onRewardedVideoShowFailed", arguments: nil)
        }
        
        func rewardedVideoDidPresent() {
            adChannel.invokeMethod("onRewardedVideoShown", arguments: nil)
        }
        
        func rewardedVideoWillDismissAndWasFullyWatched(_ wasFullyWatched: Bool) {
            adChannel.invokeMethod("onRewardedVideoClosed", arguments: ["isFinished": wasFullyWatched])
        }
        
        func rewardedVideoDidFinish(_ rewardAmount: Float, name rewardName: String?) {
            adChannel.invokeMethod("onRewardedVideoFinished",
                                   arguments: ["amount": rewardAmount, "reward": rewardName ?? "empty"])
        }
        
        func rewardedVideoDidClick() {
            adChannel.invokeMethod("onRewardedVideoClicked", arguments: nil)
        }
        
        func rewardedVideoDidExpired() {
            adChannel.invokeMethod("onRewardedVideoExpired", arguments: nil)
        }
    }
}

File: android/src/main/kotlin/com/appodeal/appodeal_flutter/AppodealRewarded.kt
package com.appodeal.appodeal_flutter

import com.appodeal.ads.RewardedVideoCallbacks
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

internal class AppodealRewarded(
    private val flutterPluginBinding: FlutterPlugin.FlutterPluginBinding
) : MethodChannel.MethodCallHandler {

    val adChannel: MethodChannel by lazy {
        MethodChannel(
            flutterPluginBinding.binaryMessenger,
            "appodeal_flutter/rewarded"
        ).apply { setMethodCallHandler(this@AppodealRewarded) }
    }

    val adListener: Listener = Listener(adChannel)

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {}

    internal class Listener(private val adChannel: MethodChannel) : RewardedVideoCallbacks {

        override fun onRewardedVideoLoaded(isPrecache: Boolean) {
            adChannel.invokeMethod("onRewardedVideoLoaded", mapOf("isPrecache" to isPrecache))
        }

        override fun onRewardedVideoFailedToLoad() {
            adChannel.invokeMethod("onRewardedVideoFailedToLoad", null)
        }

        override fun onRewardedVideoShown() {
            adChannel.invokeMethod("onRewardedVideoShown", null)
        }

        override fun onRewardedVideoShowFailed() {
            adChannel.invokeMethod("onRewardedVideoShowFailed", null)
        }

        override fun onRewardedVideoFinished(amount: Double, currency: String) {
            adChannel.invokeMethod(
                "onRewardedVideoFinished",
                mapOf("amount" to amount, "reward" to currency)
            )
        }

        override fun onRewardedVideoClosed(finished: Boolean) {
            adChannel.invokeMethod("onRewardedVideoClosed", mapOf("isFinished" to finished))
        }

        override fun onRewardedVideoExpired() {
            adChannel.invokeMethod("onRewardedVideoExpired", null)
        }

        override fun onRewardedVideoClicked() {
            adChannel.invokeMethod("onRewardedVideoClicked", null)
        }
    }
}